### ISSUE TITLE
Strange crash in SEP

### ISSUE THREAD

**Issue by cmccully** [kbarbary/sep#56]
I got this crash running sep.

*** Error in `/opt/conda/bin/python': double free or corruption (!prev): 0x00000000026cb230 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x7c503)[0x7f345a19a503]
/opt/conda/lib/python3.6/site-packages/sep.cpython-36m-x86_64-linux-gnu.so(sep_extract+0x546)[0x7f3441d37cd6]
/opt/conda/lib/python3.6/site-packages/sep.cpython-36m-x86_64-linux-gnu.so(+0x33803)[0x7f3441cf8803]
/opt/conda/lib/python3.6/site-packages/sep.cpython-36m-x86_64-linux-gnu.so(+0x44a7a)[0x7f3441d09a7a]

Looks like it is maybe overstepping the array in sep somewhere?

**Comment by kbarbary** [comment]
Could be. Or maybe with whatever set of options you're using in `extract`, a double-`free()` occurs. I'd probably have to do some print-statement debugging.

**Comment by cmccully** [comment]
Would you like me to send you the data and the command I'm running? I am using slightly non-standard parameters because I am using this for very defocused images.

**Comment by cmccully** [comment]
[bpl1m002-kb05-20180308-0259-x00.fits.gz](https://github.com/kbarbary/sep/files/1835642/bpl1m002-kb05-20180308-0259-x00.fits.gz)

This the snippit I use to run sep. 
def run_sep(data, mask_threshold=None):
    if mask_threshold is None:
        mask_threshold = 20.0 * np.sqrt(np.median(data)) + np.median(data)
    background = sep.Background(np.ascontiguousarray(data), mask=np.ascontiguousarray(data > mask_threshold))
    return sep.extract(data - background, 20.0, err=np.sqrt(data), minarea=1000, deblend_cont=1.0, filter_kernel=None)

I noticed that on my Mac this produced a set_extract_pixstack error so it might be related to that. On my linux box it produced the error above. 
Thanks!

**Comment by mkostrun** [comment]
I am rlab developer, and I have encountered this problem as well. In my case the problem was solved as following.
In file "extract.c" the following needs to be removed, because these release resources of the structure im, which was created outside the routine sep_extract. 
I propose these changes to be put in sep.

// RLAB owns these arrays: cannot be deleted by sep. Bad sep!
//   if (image->noise)
//     arraybuffer_free(&nbuf);
//   if (image->mask)
//     arraybuffer_free(&mbuf);

Cheers,
mj

**Comment by mkostrun** [comment]
Hi,
I am developer of rlabplus, for which I have implemented SEP. I have had similar problem, which in my case occured in extract function, when noise matrix was provided but 
  im.noise_type = SEP_NOISE_NONE;
  thresh_type = SEP_THRESH_ABS; 

What I think is the solution is that freeing the pointer cdscan is not done right. In line 743 of the version 1.0.0 of the file extract.c, additional check has to be implemented before freeing cdscan:

l.743: if ((cdscan != NULL) && (cdscan != dummyscan) && (cdscan != dbuf.midline))
l.744:   free(cdscan);

Hope this helps,
Marijan

**Comment by kbarbary** [comment]
Thanks @mkostrun ! I'll hopefully have a chance to try this fix out tonight.

**Comment by kbarbary** [comment]
I changed line 743 to 
```
      if (conv && (cdscan != dummyscan))
```
which I think should be right: `cdscan` is allocated if and only if `conv` is true, so it only needs to be freed if `conv` is true. But if it is equal to `dummyscan`, then we already freed it. Does that make sense?

This fixed Curtis's example. Well, it fixed the core dump. (The example still errors, but the error is handled correctly.)
